executor/cook/progress.py

```diff
@@ -0,0 +1,207 @@
+        :param force_send: boolean, optional
+            Defaults to false.
+            
+        Returns
+        -------
+        :return: Nothing 
+        """
+        with self.lock:
+            if progress_data is not None and self.last_progress_data != progress_data:
+                if force_send or self.has_enough_time_elapsed_since_last_update():
+                    logging.info('Sending progress message {}'.format(progress_data))
+                    message_dict = dict(progress_data)
+                    message_dict['task-id'] = task_id
+                    progress_message = json.dumps(message_dict)
+                    send_message(driver, progress_message, max_message_length)
+                    self.last_reported_time = time.time()
+                    self.last_progress_data = progress_data
+                else:
+                    logging.debug('Not sending progress data as enough time has not elapsed since last update')
+
+
+class ProgressWatcher(object):
+    """This class tails the output from the target file listening for progress messages.
+    The retrieve_progress_states generates all progress messages iteratively.
+    """
+
+    @staticmethod
+    def match_progress_update(progress_regex_str, input_string):
+        """Verifies whether the input string matches the provided regex.
+        
+        Parameters
+        ----------
+        :param progress_regex_str: string
+            The progress regex to match against, it must return two capture groups.
+        :param input_string: string
+            The input string.
+        
+        Returns
+        -------
+        :return: whether the input string matches the provided regex. 
+        """
+        matches = re.findall(progress_regex_str, input_string)
+        return matches[0] if len(matches) >= 1 else None
+
+    def __init__(self, config, completed_signal):
+        self.target_file = config.progress_output_name
+        self.progress_regex_string = config.progress_regex_string
+        self.progress = None
+        self.completed_signal = completed_signal
+        self.max_bytes_read_per_line = config.max_bytes_read_per_line
+
+    def current_progress(self):
+        """Returns the current progress dictionary."""
+        return self.progress
+
+    def tail(self, sleep_time_ms):
+        """This method incrementally generates lines from a file by waiting for new content from a file.
+        It behaves like the 'tail -f' shell command.
+        
+        Parameters
+        ----------
+        :param sleep_time_ms: int
+            The unit of time in ms to repetitively sleep when the file has not been created or no new
+            content is available in the file being tailed.
+        
+        Returns
+        -------
+        :return: an incrementally generated list of lines in the file being tailed.
+        """
+        try:
+            sleep_param = sleep_time_ms / 1000
+            while True:
```

Comment:
Can this be re-written as:

```python
while not os.path.isFile(self.target_file) and not self.completed_signal.isSet():
   logging.debug('{} has not yet been created, sleeping {} ms'.format(self.target_file, sleep_time_ms))
   time.sleep(sleep_param)

if self.completed_signal.isSet():
   logging.info('No output has been read to parse progress messages')

if os.path.isfile(self.target_file):
   logging.info('{} has been created, reading contents'.format(self.target_file))
```

I don't like while true as it hides the intent of the loop